File: targets/python-flask/services/order_service.py
"""Order / wallet logic."""
from data import db

BALANCES = {"alice": 1000.0, "bob": 1000.0}
USED_COUPONS = set()


def transfer(src, dst, amount):
    """Transfer amount from src to dst."""
    src_balance = BALANCES.get(src, 0.0)
    BALANCES[src] = src_balance - amount
    BALANCES[dst] = BALANCES.get(dst, 0.0) + amount
    return BALANCES[src]


def apply_coupon(user, coupon):
    """Apply a coupon credit for a user."""
    if coupon == "SAVE50":
        BALANCES[user] = BALANCES.get(user, 0.0) + 50.0
        return True
    return False


def withdraw(user, amount):
    """Withdraw amount from a user's balance."""
    balance = BALANCES.get(user, 0.0)
    if balance >= amount:
        new_balance = balance - amount
        db.execute_unsafe("UPDATE balances SET amount = %s WHERE user = '%s'" % (new_balance, user))
        BALANCES[user] = new_balance
        return True
    return False


import threading

_lock = threading.Lock()


def withdraw_safe(user, amount):
    with _lock:
        balance = BALANCES.get(user, 0.0)
        if balance >= amount:
            BALANCES[user] = balance - amount
            return True
    return False

File: targets/php-laravel/app/Services/OrderService.php
<?php

namespace App\Services;

use App\Repositories\Db;

class OrderService
{
    private static array $balances = ['alice' => 1000.0, 'bob' => 1000.0];

    public function transfer(string $src, string $dst, float $amount): float
    {
        $srcBalance = self::$balances[$src] ?? 0.0;
        self::$balances[$src] = $srcBalance - $amount;
        self::$balances[$dst] = (self::$balances[$dst] ?? 0.0) + $amount;
        return self::$balances[$src];
    }

    public function applyCoupon(string $user, string $coupon): bool
    {
        if ($coupon === 'SAVE50') {
            self::$balances[$user] = (self::$balances[$user] ?? 0.0) + 50.0;
            return true;
        }
        return false;
    }

    public function withdraw(string $user, float $amount): bool
    {
        $balance = self::$balances[$user] ?? 0.0;
        if ($balance >= $amount) {
            $newBalance = $balance - $amount;
            Db::execute("UPDATE balances SET amount = " . $newBalance . " WHERE user = '" . $user . "'");
            self::$balances[$user] = $newBalance;
            return true;
        }
        return false;
    }

    public function withdrawLocked(string $user, float $amount): bool
    {
        $lock = fopen('/tmp/orderdemo.lock', 'c');
        flock($lock, LOCK_EX);
        try {
            $balance = self::$balances[$user] ?? 0.0;
            if ($balance >= $amount) {
                self::$balances[$user] = $balance - $amount;
                return true;
            }
            return false;
        } finally {
            flock($lock, LOCK_UN);
            fclose($lock);
        }
    }
}
